File: learnLeetCode_remote/Day24/Part2/ciphered_crime_logs.py
# Ciphered Crime Logs
# Transform by XOR with k on ASCII, then return longest palindromic substring of decoded string.
# Note: The example mapping to "level" may not match simple ASCII XOR; we implement the general method per hint.
from typing import Tuple

def xor_decode(s: str, k: int) -> str:
    return ''.join(chr(ord(c) ^ k) for c in s)

def longest_palindrome(s: str) -> str:
    if not s: return s
    n = len(s)
    start, best = 0, 1
    def expand(l: int, r: int):
        nonlocal start, best
        while l>=0 and r<n and s[l]==s[r]:
            l -= 1; r += 1
        # now s[l+1:r] is palindrome
        if r - (l+1) > best:
            best = r - (l+1)
            start = l+1
    for i in range(n):
        expand(i, i)
        expand(i, i+1)
    return s[start:start+best]

def ciphered_crime_logs(s: str, k: int) -> str:
    dec = xor_decode(s, k)
    return longest_palindrome(dec)

if __name__ == "__main__":
    print(ciphered_crime_logs("jyfwf", 7))

File: Day24/Part2/ciphered_crime_logs.cpp
#include <bits/stdc++.h>
using namespace std;

// Ciphered Crime Logs: ASCII XOR decode then longest palindromic substring via expand-around-center.

string xor_decode(const string& s, int k){
    string t; t.reserve(s.size());
    for(char c: s) t.push_back(char((unsigned char)c ^ (unsigned char)k));
    return t;
}

string longest_palindrome(const string& s){
    int n = (int)s.size(); if(n==0) return string();
    int best=1, start=0;
    auto expand = [&](int l, int r){
        while(l>=0 && r<n && s[l]==s[r]){ --l; ++r; }
        int len = r-(l+1);
        if(len > best){ best=len; start=l+1; }
    };
    for(int i=0;i<n;++i){ expand(i,i); expand(i,i+1); }
    return s.substr(start, best);
}

string ciphered_crime_logs(const string& s, int k){
    string dec = xor_decode(s,k);
    return longest_palindrome(dec);
}

int main(){
    cout << ciphered_crime_logs("jyfwf", 7) << "\n";
    return 0;
}
